File: Ejercicio 16/Ejercicio 16.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>

using namespace std;

// Estructura del nodo
struct Nodo {
    int dato;
    Nodo* siguiente;
    Nodo* anterior;
    
    Nodo(int valor) {
        dato = valor;
        siguiente = nullptr;
        anterior = nullptr;
    }
};

// Clase de Lista Doblemente Enlazada
class ListaDoble {
private:
    Nodo* cabeza;
    Nodo* cola;
    
public:
    // Constructor
    ListaDoble() {
        cabeza = nullptr;
        cola = nullptr;
    }
    
    // Método para insertar al final
    void insertarFinal(int valor) {
        Nodo* nuevoNodo = new Nodo(valor);
        
        if (cabeza == nullptr) {
            // Lista vacía
            cabeza = nuevoNodo;
            cola = nuevoNodo;
        } else {
            // Insertar al final
            cola->siguiente = nuevoNodo;
            nuevoNodo->anterior = cola;
            cola = nuevoNodo;
        }
    }
    
    // Método para mostrar la lista de inicio a fin
    void mostrarAdelante() {
        if (cabeza == nullptr) {
            cout << "La lista está vacía." << endl;
            return;
        }
        
        Nodo* actual = cabeza;
        cout << "Lista (adelante): ";
        while (actual != nullptr) {
            cout << actual->dato << " ";
            actual = actual->siguiente;
        }
        cout << endl;
    }
    
    // Método para mostrar la lista de fin a inicio
    void mostrarAtras() {
        if (cola == nullptr) {
            cout << "La lista está vacía." << endl;
            return;
        }
        
        Nodo* actual = cola;
        cout << "Lista (atrás): ";
        while (actual != nullptr) {
            cout << actual->dato << " ";
            actual = actual->anterior;
        }
        cout << endl;
    }
    
    // Método para generar lista con valores aleatorios
    void generarListaAleatoria(int cantidad) {
        // Semilla para números aleatorios
        srand(time(0));
        
        for (int i = 0; i < cantidad; i++) {
            int valor = rand() % 100 + 1; // Números entre 1 y 100
            insertarFinal(valor);
        }
    }
    
    // Destructor para liberar memoria
    ~ListaDoble() {
        Nodo* actual = cabeza;
        while (actual != nullptr) {
            Nodo* temp = actual;
            actual = actual->siguiente;
            delete temp;
        }
    }
};

// Función principal
int main() {
    ListaDoble lista;
    
    cout << "=== LISTA DOBLEMENTE ENLAZADA CON VALORES ALEATORIOS ===" << endl;
    
    // Generar lista con 15 valores aleatorios
    lista.generarListaAleatoria(15);
    
    // Mostrar la lista en ambas direcciones
    lista.mostrarAdelante();
    lista.mostrarAtras();
    
    return 0;
}

File: Ejercicio 16/Ejercicio 16.java
import java.util.Random;

// Clase del nodo
class Nodo {
    public int dato;
    public Nodo siguiente;
    public Nodo anterior;
    
    public Nodo(int valor) {
        dato = valor;
        siguiente = null;
        anterior = null;
    }
}

// Clase de Lista Doblemente Enlazada
class ListaDoble {
    private Nodo cabeza;
    private Nodo cola;
    
    // Constructor
    public ListaDoble() {
        cabeza = null;
        cola = null;
    }
    
    // Método para insertar al final
    public void insertarFinal(int valor) {
        Nodo nuevoNodo = new Nodo(valor);
        
        if (cabeza == null) {
            // Lista vacía
            cabeza = nuevoNodo;
            cola = nuevoNodo;
        } else {
            // Insertar al final
            cola.siguiente = nuevoNodo;
            nuevoNodo.anterior = cola;
            cola = nuevoNodo;
        }
    }
    
    // Método para mostrar la lista de inicio a fin
    public void mostrarAdelante() {
        if (cabeza == null) {
            System.out.println("La lista está vacía.");
            return;
        }
        
        Nodo actual = cabeza;
        System.out.print("Lista (adelante): ");
        while (actual != null) {
            System.out.print(actual.dato + " ");
            actual = actual.siguiente;
        }
        System.out.println();
    }
    
    // Método para mostrar la lista de fin a inicio
    public void mostrarAtras() {
        if (cola == null) {
            System.out.println("La lista está vacía.");
            return;
        }
        
        Nodo actual = cola;
        System.out.print("Lista (atrás): ");
        while (actual != null) {
            System.out.print(actual.dato + " ");
            actual = actual.anterior;
        }
        System.out.println();
    }
    
    // Método para generar lista con valores aleatorios
    public void generarListaAleatoria(int cantidad) {
        Random random = new Random();
        
        for (int i = 0; i < cantidad; i++) {
            int valor = random.nextInt(100) + 1; // Números entre 1 y 100
            insertarFinal(valor);
        }
    }
}

// Clase principal
public class Main {
    public static void main(String[] args) {
        ListaDoble lista = new ListaDoble();
        
        System.out.println("=== LISTA DOBLEMENTE ENLAZADA CON VALORES ALEATORIOS ===");
        
        // Generar lista con 15 valores aleatorios
        lista.generarListaAleatoria(15);
        
        // Mostrar la lista en ambas direcciones
        lista.mostrarAdelante();
        lista.mostrarAtras();
    }
}